lib/core/base/audit.js

```diff
@@ -327,58 +436,79 @@ Audit.prototype.addCheck = function(spec) {
+
 	this.normalizeOptions(options);
 
 	axe._selectCache = [];
-	var q = axe.utils.queue();
-	this.rules.forEach(function(rule) {
-		if (axe.utils.ruleShouldRun(rule, context, options)) {
-			if (options.performanceTimer) {
-				var markEnd = 'mark_rule_end_' + rule.id;
-				var markStart = 'mark_rule_start_' + rule.id;
-				axe.utils.performanceTimer.mark(markStart);
+
+	// get a list of rules to run NOW vs. LATER (later are preload assets dependent rules)
+	const allRulesToRun = getRulesToRun(this.rules, context, options); //TODO:JEY - test
+	const runNowRules = allRulesToRun.now;
+	const runLaterRules = allRulesToRun.later;
+
+	// init a NOW queue for rules to run immediately
+	const nowQueue = axe.utils.queue();
+
+	// construct can run NOW rules into NOW queue
+	runNowRules.forEach(rule => {
+		const deferredRule = getDefferedRule(rule, context, options);
+		nowQueue.defer(deferredRule);
+	});
+
+	// defer preload into the NOW queue
+	// only if there are rules that need assets
+	if (runLaterRules.length) {
+		const preloader = axe.utils.preload(options);
+		nowQueue.defer(preloader);
+	}
+
+	// invoke the now queue
+	nowQueue
+		.then(nowQueueResults => {
+			// if there are no rules to run LATER - resolve with rule results
+			if (!runLaterRules.length) {
+				// remove the cache
+				axe._selectCache = undefined;
+
+				// resolve
+				resolve(nowQueueResults.filter(result => !!result));
+				return;
 			}
-			q.defer(function(res, rej) {
-				rule.run(
+
+			// if there are rules to be run LATER based on assets
+			// interpolate results into separate variables
+			// as preloader was the last deferred,
+			// the last item in the results is the preloaded assets (hence pop())
+			// extract first item in array as `queue` returns an array
+			const preloadedAssets = nowQueueResults.pop()[0];
+
+			// the reminder of the results are RuleResults
+			const nowRulesResults = nowQueueResults;
+
+			// init a LATER queue for rules that are dependant on preloaded assets
+			const laterQueue = axe.utils.queue();
+			runLaterRules.forEach(rule => {
+				const deferredRule = getDefferedRule(
+					rule,
 					context,
 					options,
-					function(out) {
-						if (options.performanceTimer) {
-							axe.utils.performanceTimer.mark(markEnd);
-							axe.utils.performanceTimer.measure(
-								'rule_' + rule.id,
-								markStart,
-								markEnd
-							);
-						}
-						res(out);
-					},
-					function(err) {
-						if (!options.debug) {
-							var errResult = Object.assign(new RuleResult(rule), {
-								result: axe.constants.CANTTELL,
-								description: 'An error occured while running this rule',
-								message: err.message,
-								stack: err.stack,
-								error: err
-							});
-							res(errResult);
-						} else {
-							rej(err);
-						}
-					}
+					preloadedAssets
 				);
+				laterQueue.defer(deferredRule);
 			});
-		}
-	});
-	q.then(function(results) {
-		axe._selectCache = undefined; // remove the cache
```

Comment:
I have my doubts about if these changes clear the selector cache properly. Please make sure there are tests showing that it does for all paths through this code.

Reply:
Tests added for the same, for various code paths.